src/main/kotlin/org/rust/ide/typing/paste/RsConvertJsonToStructCopyPasteProcessor.kt

```diff
@@ -0,0 +1,199 @@
+            project,
+            icon = Messages.getQuestionIcon()
+        )
+    }
+}
+
+private fun insertStruct(
+    caretOffset: Int,
+    editor: Editor,
+    bounds: RangeMarker,
+    struct: RsStructItem
+) {
+    val psiDocumentManager = PsiDocumentManager.getInstance(struct.project)
+
+    runWriteAction {
+        // Delete original text
+        editor.document.deleteString(bounds.startOffset, bounds.endOffset)
+        psiDocumentManager.commitDocument(editor.document)
+
+        val file = psiDocumentManager.getPsiFile(editor.document) as? RsFile ?: return@runWriteAction
+        val element = file.findElementAt(caretOffset)
+        val inserted = if (element == null) {
+            file.add(struct)
+        } else {
+            element.parent.addAfter(struct, element)
+        } as RsStructItem
+
+        val name = inserted.identifier?.createSmartPointer()
+        val wildCards = inserted.descendantsOfType<RsPatWild>().map { it.createSmartPointer() }
+
+        // TODO: make template work :(
+        editor.buildAndRunTemplate(inserted.containingFile, listOfNotNull(name) + wildCards)
+    }
+}
+
+private class PotentialJsonTransferableData(val text: String) : TextBlockTransferableData {
+    override fun getFlavor() = DATA_FLAVOR
+    override fun getOffsetCount() = 0
+
+    override fun getOffsets(offsets: IntArray?, index: Int) = index
+    override fun setOffsets(offsets: IntArray?, index: Int) = index
+
+    companion object {
+        val DATA_FLAVOR: DataFlavor = DataFlavor(
+            RsConvertJsonToStructCopyPasteProcessor::class.java,
+            "class: RsConvertJsonToStructCopyPasteProcessor"
+        )
+    }
+}
+
+private fun generateStruct(jsonFile: JsonFile): String? {
+    val rootObject = jsonFile.topLevelValue as? JsonObject ?: return null
+
+    return buildString {
+        append("struct Struct {\n")
+
+        for (property in rootObject.propertyList) {
+            val value = property.value ?: continue
+            val serdeType = getSerdeType(value)
+            append("${property.name}: ${serdeType},\n")
+        }
+
+        append("}")
+    }
+}
+
+private fun getSerdeType(value: JsonValue): String {
+    val atomic = getAtomicSerdeType(value)
+    if (atomic != null) {
+        return atomic
+    }
+    return when (value) {
+        is JsonArray -> {
+            val values = value.valueList
+            if (values.isEmpty()) {
+                "Vec<_>"
+            } else {
+                val first = values[0]
+                val subtype = getAtomicSerdeType(first) ?: "_"
+                "Vec<$subtype>"
+            }
+        }
+        else -> "_"
+    }
+}
+
+private fun getAtomicSerdeType(value: JsonValue): String? {
+    return when (value) {
+        is JsonBooleanLiteral -> "bool"
+        is JsonStringLiteral -> "String"
+        is JsonNumberLiteral -> {
+            val text = value.text
+            when {
+                text.toBigIntegerOrNull() != null -> "u64"
```

Comment:
Good point. `i64`/`i32` is more general I suppose. Maybe we could special-case so that `i64` is generated only if the number is negative. But that seems to be too much magic. I'll switch it to `i64`/`i32`.